### ISSUE TITLE
Create a public-api-only core dependency

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1156]
Provide a 

### Expected Behavior
Allow a user to declare a dependency on only the `gradlePublicApi()`. 

### Current Behavior
Currently, users wanting to write plugins must declare a `compile gradleApi()` dependency which exposes all of Gradle's internals and pulls in a bunch of transitive dependencies. 

### Context
Plugin authors are not prevented from using Gradle internal APIs (which are sadly necessary for some users' needs today) which breaks backward and forward compatibility. 

Users have complained about this situation [again](https://discuss.gradle.org/t/implementing-a-custom-task-without-depending-on-gradles-internals/19530), and [again](http://markmail.org/message/ugsgvsazykbeq7hx), and [again](http://stackoverflow.com/questions/33906081/gradle-custom-plugin-gradleapi-vs-explicit-dependency).

Let us give them a vehicle to ensure compatibility, then [make that vehicle awesome](https://github.com/gradle/gradle/issues/716).

**Comment by adammurdoch** [comment]
We should bust up the API into the public API for the Gradle platform, and the public API of each group of plugins. For example:

```dependencies {
    compile gradlePublicApi() // only the API of the Gradle platform, no plugins
    compile plugin id 'java'  // the API of the java plugin
    compile plugin id 'my.org.plugin' version '1.2' // the API of my custom plugin
}```

**Comment by bmuschko** [comment]
Initial spec: https://github.com/gradle/gradle/blob/master/design-docs/internal-apis-made-public.md#story-dependencyhandler-exposes-methods-for-referring-to-public-and-internal-gradle-api

**Comment by lacasseio** [comment]
Issues like https://github.com/gradle/gradle/pull/1021 may need to be resolved in order to give good user experience.

**Comment by bmuschko** [comment]
@lacasseio I don't think it should necessarily hold us back from implementing this issue but to be able to fully avoid the use of internal APIs #1021 would have to be fixed. Some people might not be affected e.g. if you implement your logic in Groovy and don't use strong typing.

**Comment by lacasseio** [comment]
@bmuschko You are right. It's something to keep on the radar.

**Comment by adammurdoch** [comment]
Just repeating a previous comment: We should bust up the API into the public API for the Gradle platform, and the public API of each group of plugins.

**Comment by bmuschko** [comment]
@adammurdoch Can you have a look at the spec I wrote as a first step of approaching the problem? I'd like to defer implementing a public API for the plugins in a later story.

**Comment by JLLeitschuh** [comment]
@bmuschko The link to the spec you wrote is broken.

**Comment by bmuschko** [comment]
@JLLeitschuh AFAIR the `design-docs` folder has been deleted from the repository. I am not quite sure if there is a replacement for it.

**Comment by JLLeitschuh** [comment]
@bmuschko Do you want to post the proposal you had somewhere where it can still be read?

**Comment by bmuschko** [comment]
@JLLeitschuh I don't have it anymore on my local machine. I am sure you can find it in the Git history though for an older tag.

**Comment by JLLeitschuh** [comment]
Dug it out:
https://github.com/gradle/gradle/blob/774d412706877e10319addbfb675d1f27343f5ed/design-docs/internal-apis-made-public.md#story-dependencyhandler-exposes-methods-for-declaring-public-and-internal-gradle-api

**Comment by TWiStErRob** [comment]
Are you planning to do this for 5?

**Comment by LDVSOFT** [comment]
Any progress here? It's been quite a while.

**Comment by TWiStErRob** [comment]
@lptr Are you planning to do this for 8? :)

It would help clarifying the needs of Gradle plugin authors, and defining a more useful public interface so we don't need to go into the internals.

**Comment by lptr** [comment]
We are planning some improvements in this area, but there is no clear timeline yet.

**Comment by liutikas** [comment]
Are there any updates on this? gradle-api.jar just keeps getting bigger and bigger with every release with more and more internal implementation classes showing up on the classpath. The generation of this jar is 6s seconds, it is a footgun making it easy to pull in classes you should not use, such as `org.gradle.internal.impldep.com.google.common.collect.ImmutableMap`.

In the meantime, for plugin authors, my team started shipping a lint check to help catch these accidental uses since Gradle has been really slow to respond to this.

```
plugins {
  id("com.android.lint")
}
dependencies {
  lintChecks("androidx.lint:lint-gradle:1.0.0-alpha01")
}
```

this jar exists on maven.google.com
